sled-agent/src/opte.rs

```diff
@@ -0,0 +1,334 @@
+}
+
+impl Default for BoundaryServices {
+    fn default() -> Self {
+        // TODO-completeness: Don't hardcode this.
+        //
+        // Boundary Services will be started on several sleds during rack
+        // setup, and the addresses of the chosen sleds will need to be
+        // propagated here.
+        const BOUNDARY_SERVICES_ADDR: Ipv6Addr =
+            Ipv6Addr::new(0xfd00, 0x99, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01);
+        let boundary_services_vni = Vni::new(99_u32).unwrap();
+
+        Self { ip: BOUNDARY_SERVICES_ADDR, vni: boundary_services_vni }
+    }
+}
+
+/// Information about the gateway for an OPTE port
+#[derive(Debug, Clone, Copy)]
+pub struct Gateway {
+    mac: MacAddr6,
+    ip: IpAddr,
+}
+
+impl Gateway {
+    pub fn from_subnet(subnet: &IpNetwork) -> Self {
+        Self {
+            // TODO-completeness: We need a robust way of assigning a
+            // non-clashing MAC address for the virtual gateway that OPTE
+            // represents.
+            mac: MacAddr6::new(0xa8, 0x40, 0x25, 0x00, 0x00, 0x01),
+
+            // See RFD 21, section 2.2 table 1
+            ip: subnet
+                .iter()
+                .nth(1)
+                .expect("IP subnet must have at least 1 address"),
+        }
+    }
+}
+
+/// A port on the OPTE "virtual switch", which corresponds to one guest network
+/// interface.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct OptePort {
+    name: String,
+    ip: IpAddr,
+    subnet: IpNetwork,
+    mac: MacAddr6,
+    vni: Vni,
+    underlay_ip: Ipv6Addr,
+    gateway: Gateway,
+    boundary_services: BoundaryServices,
+    // TODO-correctness: Remove this once we can put Viona directly on top of an
+    // OPTE port device.
+    //
+    // Note that this will always be `Some(_)`. It is wrapped in an optional to
+    // ensure we can drop the VNIC before we drop the OPTE port itself.
+    vnic: Option<Vnic>,
+}
+
+impl OptePort {
+    /// Return the VNIC used to link OPTE and Viona.
+    // TODO-correctness: Remove this once we can put Viona directly on top of an
+    // OPTE port device.
+    pub fn vnic(&self) -> &Vnic {
+        self.vnic.as_ref().unwrap()
+    }
+}
+
+impl Drop for OptePort {
+    fn drop(&mut self) {
+        self.vnic.take();
+        if let Ok(hdl) = OpteHdl::open(OpteHdl::DLD_CTL) {
+            if hdl.delete_xde(&self.name).is_ok() {
+                return;
+            }
+        }
+        eprintln!("WARNING: Failed to delete OPTE port '{}'", self.name);
+    }
+}
+
+/// Initialize the underlay devices required for the XDE kernel module.
+///
+/// This needs the sled's main underlay address, to route traffic between
+/// guests on this host and those on other Gimlets, and the physical NICs over
+/// which it can send that traffic.
+pub fn initialize_xde_driver(log: &Logger) -> Result<(), Error> {
+    let underlay_nics = find_chelsio_links()?;
+    info!(log, "using '{:?}' as data links for XDE driver", underlay_nics);
```

Comment:
Any reason to capitalize `xde`? AFAIK it's not an acronym. Typically driver names are just referenced in lowercase.

Reply:
Not particularly, just that OPTE is capitalized. That's an acronym though, as you point out. I'll make a pass and lowercase everything.